From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Vasiliy Kovalev <kovalev@altlinux.org>
Date: Mon, 31 Aug 2026 10:05:00 +0000
Subject: [PATCH] sys/linux: add dm-secdel descriptions

Constructor grammar for the out-of-tree dm-secdel target (device, offset
and erase pattern) and the block I/O that reaches secdel_map().

---
 sys/linux/dev_dm_secdel.txt       | 57 +++++++++++++++++++++++++++++++
 sys/linux/dev_dm_secdel.txt.const |  7 ++++
 2 files changed, 64 insertions(+)
 create mode 100644 sys/linux/dev_dm_secdel.txt
 create mode 100644 sys/linux/dev_dm_secdel.txt.const

diff --git a/sys/linux/dev_dm_secdel.txt b/sys/linux/dev_dm_secdel.txt
new file mode 100644
index 0000000..3fae568
--- /dev/null
+++ b/sys/linux/dev_dm_secdel.txt
@@ -0,0 +1,57 @@
+include <uapi/linux/dm-ioctl.h>
+include <uapi/linux/fcntl.h>
+include <uapi/linux/fs.h>
+
+# ==================== IOCTLS / CALLS ====================
+
+openat$dm_secdel_io(fd const[AT_FDCWD], file ptr[in, string[dm_secdel_nodes]], flags flags[open_flags], mode const[0]) fd_dm
+
+# ==================== FLAGS & PARAMS ====================
+
+dm_secdel_nodes = "/dev/ram0", "/dev/dm-0", "/dev/dm-1", "/dev/dm-2", "/dev/dm-3", "/dev/dm-4", "/dev/dm-5", "/dev/mapper/syzio", "/dev/mapper/syzpart", "/dev/mapper/syzkeep", "/dev/mapper/syzsecdel0"
+
+syz_dm_setup$secdel(index intptr[0:7], target ptr[in, string["secdel"]], params ptr[in, dm_secdel_params], sectors flags[dm_target_lengths]) fd_dm
+syz_dm_setup$secdel_str(index intptr[0:7], target ptr[in, string["secdel"]], params ptr[in, string[dm_secdel_param_strings]], sectors flags[dm_target_lengths]) fd_dm
+
+ioctl$DM_TABLE_LOAD_secdel(fd fd_dm_control, cmd const[DM_TABLE_LOAD], arg ptr[inout, dm_load_secdel])
+ioctl$DM_TABLE_LOAD_secdel_str(fd fd_dm_control, cmd const[DM_TABLE_LOAD], arg ptr[inout, dm_load_secdel_str])
+
+# ==================== STRUCTS ====================
+
+type dm_load_secdel dm_ioctl_t[1, dm_ident_name, flags[dm_ioctl_flags, int32], dm_target_secdel]
+type dm_load_secdel_str dm_ioctl_t[1, dm_ident_name, flags[dm_ioctl_flags, int32], dm_target_secdel_str]
+
+dm_target_secdel {
+	sector_start	const[0, int64]
+	length		flags[dm_target_lengths, int64]
+	status		const[0, int32]
+	next		len[parent, int32]
+	target_type	string["secdel", DM_MAX_TYPE_NAME]
+	params		dm_secdel_params
+} [packed]
+
+dm_secdel_params {
+	dev	stringnoz[dm_secdel_devices]
+	sp0	const[' ', int8]
+	start	fmt[dec, flags[dm_secdel_starts]]
+	sp1	const[' ', int8]
+	pattern	array[flags[dm_secdel_pattern_chars, int8], 0:8]
+	nul	const[0, int8]
+} [packed]
+
+dm_secdel_starts = 0, 8, 4096, 32768, 65535, 65536, 18446744073709551615
+
+dm_secdel_devices = "1:0", "1:1", "1:2", "1:3", "/dev/ram0", "/dev/ram1", "7:0", "/dev/loop0", "0:0", "259:0", "/dev/does-not-exist"
+
+dm_secdel_pattern_chars = 0x30, 0x31, 0x52, 0x32, 0x72, 0x00
+
+dm_target_secdel_str {
+	sector_start	const[0, int64]
+	length		flags[dm_target_lengths, int64]
+	status		const[0, int32]
+	next		len[parent, int32]
+	target_type	string["secdel", DM_MAX_TYPE_NAME]
+	params		string[dm_secdel_param_strings, 48]
+}
+
+dm_secdel_param_strings = "1:0 0 R", "1:0 0 1R0", "1:0 0 R1R0R1R0", "1:0 0", "1:0", "1:0 x R", "1:0 -1 R", "1:0 18446744073709551615 R", "1:0 0 XYZ", "1:0 0 \R", "1:0 0 'R'", "1:0 0 R extra", "1:0 0 R 1 2 3 4 5 6 7 8", "", "1:0 0 RRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRR"
diff --git a/sys/linux/dev_dm_secdel.txt.const b/sys/linux/dev_dm_secdel.txt.const
new file mode 100644
index 0000000..59907e9
--- /dev/null
+++ b/sys/linux/dev_dm_secdel.txt.const
@@ -0,0 +1,7 @@
+# Code generated by syz-sysgen. DO NOT EDIT.
+arches = amd64
+AT_FDCWD = amd64:18446744073709551516
+DM_MAX_TYPE_NAME = amd64:16
+DM_TABLE_LOAD = amd64:3241737481
+__NR_ioctl = amd64:16
+__NR_openat = amd64:257
